renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -117,10 +118,211 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
+                element.type == eRootTable)
+        {
+          UINT prevTableOffset = 0;
+          WrappedID3D12DescriptorHeap *heap =
+              rm->GetCurrentAs<WrappedID3D12DescriptorHeap>(element.id);
+
+          size_t numRanges = param.ranges.size();
+          for(size_t r = 0; r < numRanges; ++r)
+          {
+            const D3D12_DESCRIPTOR_RANGE1 &range = param.ranges[r];
+
+            UINT offset = range.OffsetInDescriptorsFromTableStart;
+            if(range.OffsetInDescriptorsFromTableStart == D3D12_DESCRIPTOR_RANGE_OFFSET_APPEND)
+              offset = prevTableOffset;
+
+            D3D12Descriptor *desc = (D3D12Descriptor *)heap->GetCPUDescriptorHandleForHeapStart().ptr;
+            desc += element.offset;
+            desc += offset;
+
+            UINT numDescriptors = range.NumDescriptors;
+            if(numDescriptors == UINT_MAX)
+            {
+              // Find out how many descriptors are left after
+              UINT availDescriptors = heap->GetNumDescriptors() - offset - (UINT)element.offset;
+
+              // TODO: Should we look up the bind point in the D3D12 state to try to get
+              // a better guess at the number of descriptors?
+
+              numDescriptors = RDCMIN(availDescriptors, 128U);
+            }
+
+            prevTableOffset = offset + numDescriptors;
+
+            UINT shaderReg = range.BaseShaderRegister;
+
+            if(range.RangeType == D3D12_DESCRIPTOR_RANGE_TYPE_SRV)
+            {
+              for(UINT n = 0; n < numDescriptors; ++n, ++shaderReg)
+              {
+                if(desc)
+                {
+                  ResourceId srvId = desc->GetResResourceId();
+                  ID3D12Resource *pResource = rm->GetCurrentAs<ID3D12Resource>(srvId);
+
+                  D3D12_SHADER_RESOURCE_VIEW_DESC srvDesc = desc->GetSRV();
+                  if(srvDesc.Format != DXGI_FORMAT_UNKNOWN)
+                  {
+                    ShaderDebug::FillViewFmt(srvDesc.Format, global.srvs[shaderReg].format);
+                  }
+                  else
+                  {
+                    D3D12_RESOURCE_DESC resDesc = pResource->GetDesc();
+                    if(resDesc.Dimension == D3D12_RESOURCE_DIMENSION_BUFFER)
+                    {
+                      global.srvs[shaderReg].format.stride = srvDesc.Buffer.StructureByteStride;
+
+                      // If we didn't get a type from the SRV description, try to pull it from the
+                      // shader reflection info
+                      ShaderDebug::LookupSRVFormatFromShaderReflection(
+                          *dxbc->GetReflection(), (uint32_t)shaderReg, global.srvs[shaderReg].format);
+                    }
+                  }
+
+                  if(srvDesc.ViewDimension == D3D12_SRV_DIMENSION_BUFFER)
+                  {
+                    // I know this isn't what the docs say, but as best as I can tell
+                    // this is how it's used.
```

Comment:
This comment is out of date :). In D3D12 they removed the weird unions, `FirstElement` / `NumElements` are the only members now.